dask_grblas/expr.py

```diff
@@ -458,17 +555,105 @@ def __call__(self, *args, **kwargs):
+        obj = obj[::-1].new()
+        obj = obj[obj_indices].new()
+        index = unique_indices
+    return index, obj
+
+
+class Assigner:
+    def __init__(self, updater, index):
+        self.updater = updater
+        self.parent = updater.parent
+        self.index = index
+        self._meta = updater.parent._meta[index]
+
+    def update(self, obj):
+        mask = self.updater.mask
+        if mask is not None:
+            assert isinstance(mask, Mask)
+            assert self.parent._meta.nvals == 0
+            meta = self._meta.new(mask=mask._meta)
+            delayed_mask = mask.mask._delayed
+            grblas_mask_type = get_grblas_type(mask)
+        else:
+            meta = self._meta.new()
+            delayed_mask = None
+            grblas_mask_type = None
+
+        if len(self.parent.shape) == 1 and type(self.index) is not tuple:
+            if type(self.index) in {list, np.ndarray}:
+                self.index, obj = _uniquify(self.index, obj)
+            indices = (self.index,)
+        else:
+            indices = self.index
+        if len(indices) == 1 or len(indices) == 2:
+            delayed_dup = self.parent._optional_dup()
+            # Extractor level 0:
+            delayed = delayed_dup.map_blocks(
+                Extractor,
+                dtype=np_dtype(meta.dtype),
+            )
+            # Extractor level 1 (and possibly 2):
+            # We use blockwise here to enable chunk alignment between
+            # `delayed_mask` and `delayed_dup`
+            if mask is not None:
+                args = [delayed, "i", delayed_mask, "i"]
+            else:
+                args = [delayed, "i"]
+            delayed = da.core.blockwise(
+                Extractor,
+                "i",
+                *args,
+                dtype=np_dtype(meta.dtype),
+                replace=self.updater.replace,
+                mask_type=grblas_mask_type,
+            )
+            delayed = delayed[indices]
+            # the size of `delayed` is now in general different from
+            # `delayed_dup` by virtue of `indices`.
+            # The following updates the indexed elements of
+            # `delayed_dup` but returns a dask.array of zeros
+            # since it is difficult to return the updated
+            # `delayed_dup` directly:
+            delayed = da.core.elemwise(
+                _extractor_update,
+                delayed,
+                np_dtype(meta.dtype),
+                grblas_mask_type,
+                self.updater.accum,
+                obj._delayed,
+                dtype=np_dtype(meta.dtype),
+            )
+            self.parent._delayed = delayed_dup
+            # The following two lines are only meant to trigger the
+            # above update whenever self.parent.compute() is called
+            zero = get_return_type(meta)(delayed).reduce().new()
+            self.parent << self.parent.apply(gb.binary.plus, right=zero)
```

Comment:
This is a very curious piece of code.  It is here because I couldn't find a way around the following problem:
The `dask` routines that are updating the `Vector` are themselves unable to return the updated `Vector` because the editing pattern of these `dask` routines cannot be determined before runtime, since the pattern is index-dependent.

This is the only way I could think of to trigger the update when `compute()` is called on the `Vector` that is to be updated.